rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/FluxStats.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.rsocket.loadbalance;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import io.rsocket.Payload;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+class FluxStats extends Flux<Payload> implements Subscription, Subscriber<Payload> {
+
+	final Flux<Payload> upstream;
+	final Stats stats;
+
+	Subscription                    s;
+	CoreSubscriber<? super Payload> actual;
+
+	volatile int                                      state;
+	static final AtomicIntegerFieldUpdater<FluxStats> STATE =
```

Comment:
I'm not familiar with `AtomicIntegerFieldUpdater`. Why choose it over `AtomicBoolean`?

Reply:
3 states

Reply:
Basically, to reduce the number of allocation, there is no AtomicBooleanFieldUpdater or soon thus to keep the number of the allocated objects as small as possible, uses that one